Add end to end encryption blog (#6)
excalidraw/excalidraw-blog: content/blog/end-to-end-encryption/index.md

---
title: End-to-End Encryption in the Browser
date: '2020-03-21T22:40:32.169Z'

> description: "&nbsp;"

Reviewer: Why not adding a small description? It will appear on the front page.

Author: I feel like the title is good enough. I'm not sure why I need to add another sentence here

Reviewer: Fair enough.. then let's just delete the description altogether, and in the upcoming PRs fix the open graph part when it's shared.

Author: Can’t delete, it throws a graphql error :(

Reviewer: OK.. something for later.

---

> [Excalidraw](https://excalidraw.com/) is a whiteboard tool that lets you easily sketch diagrams that have a hand-drawn feed to them. It is very handy to dump your thoughts many of which are sensitive: designs for new features not yet released, interview questions, org charts...

Reviewer 2: ```suggestion
[Excalidraw](https://excalidraw.com/) is a whiteboard tool that lets you easily sketch diagrams that have a hand-drawn feel to them. It is very handy to dump your thoughts many of which are sensitive: designs for new features not yet released, interview questions, org charts...
```

> By default, [Excalidraw](https://excalidraw.com/) is not sending any of the content you are drawing to the server. But sometimes it is useful to be able to send a link to the scene you are working on to someone else.

Reviewer 2: ```suggestion
By default, [Excalidraw](https://excalidraw.com/) doesn’t send anything you draw to the server. But sometimes it is useful to be able to send a link to the scene you are working on to someone else.
```

> You could save the file, send it to the other person, have them open it. But that's pretty cumbersome. In this post we'll see how to get a link working without the server having access to the data.

Reviewer 2: ```suggestion
You could save the drawing to a file, send it to the other person, and have them open it. But that's pretty cumbersome. In this post we'll show how it’s possible to share just a link without the server having access to the data.
```

Reviewer: scene instead of drawing.

## Traditional Website Architecture

> In the traditional website architecture, if you want to save a scene, you're going to send the content to the server so that it can store it. Then to load it you'll the server to send it back to you.

Reviewer 2: ```suggestion
In a traditional website architecture, you’d save a scene by sending it to the server, which gives you a shareable URL. The recipient then downloads the scene data from the server.
```

> In that world, you trust the server to contain your information but you don't trust the pipes in-between the client and server using https to encrypt the data.

Reviewer 2: ```suggestion
In that world, you trust the server to contain your information but you don't have to trust the pipes in between the client and the server, because you use HTTPS to encrypt the data.
```

[![url](url.png)](https://excalidraw.com/#json=5660568841093120,vki3y9xuEulFVHDqt-PBMw)

## Show me the code

Fortunately, the Web Cryptography APIs are now widely available to all the browsers that let us implement this. That said, the APIs to deal with encryption, keys and binary data are not the most straightforward, this next section walks you through how to wire it all together.

### Upload

We generate a random key that will be used to encrypt the data.

```javascript
const key = await window.crypto.subtle.generateKey(
  { name: "AES-GCM", length: 128 },
  true, // extractable
  ["encrypt", "decrypt"],
);
```

We encrypt the content with that random key. In this case, we only encrypt the content once with the random key so we don't need an `iv` and can leave it filled with 0.

```javascript
const encrypted = await window.crypto.subtle.encrypt(
  { name: "AES-GCM", iv: new Uint8Array(12) /* don't reuse key! */ },
  key,
  new TextEncoder().encode(JSON.stringify(content)),
);
```

We upload the encrypted content to the server. Note that we don't send the key to the server!

```javascript

> const response = await (await fetch("upload.php", {

Reviewer: can we get rid of PHP? I'm not too religious about it, but we are not using PHP anywhere in Excalidraw.

Author: I tried to show that it's an endpoint. I guess I'm old school... I'll remove

  method: "POST",
  body: encrypted
})).json();
```

We generate the shareable URL. We use the `jwk` encoding in order to extract a base64 version of the key instead of having a binary encoded one.

```javascript
const objectURL = response.url;
const objectKey = (await window.crypto.subtle.exportKey("jwk", key)).k;
const url = objectURL + '#key=' + objectKey;
// Example: https://excalidraw.com/?scene=1234#key=BQ1moYESmTEXgtA1KozyVw
```

## Download

In the opposite direction, we download the file back from the server.

```javascript

> const response = await fetch(`download.php?id={id}`);

Reviewer: same

const encrypted = await response.arrayBuffer();
```

The key that we encoded in the url is the `k` field of the `jwk` object that represents the key. In order to get back a full key object we need to reproduce all the other fields that are static. It's pretty verbose but it works!

```javascript
const objectKey = window.location.hash.slice('#key='.length);
const key = window.crypto.subtle.importKey(
  "jwk",
  {
    k: objectKey,
    alg: "A128GCM", ext: true, key_ops: ["encrypt", "decrypt"], kty: "oct",
  },
  { name: "AES-GCM", length: 128},
  false, // extractable
  ["decrypt"],
);
```

> We decrypt the message, decode it to string and parse it back as json.

Reviewer: ```suggestion
We decrypt the message, decode it to string and parse it back as JSON.
```
